components/places/src/storage/tags.rs

```diff
@@ -0,0 +1,317 @@
+        &[(":tag", &tag), (":place_id", &place_id)],
+    )?;
+    tx.commit()?;
+    Ok(())
+}
+
+/// Remove the specified tag from the specified URL.
+///
+/// # Arguments
+///
+/// * `conn` - A database connection on which to operate.
+///
+/// * `url` - The URL from which the tag should be removed.
+///
+/// * `tag` - The tag to remove from the URL.
+///
+/// # Returns
+///
+/// There is no success return value - the operation is ignored if the URL
+/// does not have the tag.
+pub fn untag_url(conn: &Connection, url: &Url, tag: String) -> Result<()> {
+    let tag = validate_tag(&tag)?;
+    conn.execute_named_cached(
+        "DELETE FROM moz_tags_relation
+         WHERE tag_id = (SELECT id FROM moz_tags
+                         WHERE tag = :tag)
+         AND place_id = (SELECT id FROM moz_places
+                         WHERE url_hash = hash(:url)
+                         AND url = :url)",
+        &[(":tag", &tag), (":url", &url.as_str())],
+    )?;
+    Ok(())
+}
+
+/// Remove all tags from the specified URL.
+///
+/// # Arguments
+///
+/// * `conn` - A database connection on which to operate.
+///
+/// * `url` - The URL for which all tags should be removed.
+///
+/// # Returns
+///
+/// There is no success return value.
+pub fn remove_all_tags_from_url(conn: &Connection, url: &Url) -> Result<()> {
+    conn.execute_named_cached(
+        "DELETE FROM moz_tags_relation
+         WHERE
+         place_id = (SELECT id FROM moz_places
+                     WHERE url_hash = hash(:url)
+                     AND url = :url)",
+        &[(":url", &url.as_str())],
+    )?;
+    Ok(())
+}
+
+/// Remove the specified tag from all URLs.
+///
+/// # Arguments
+///
+/// * `conn` - A database connection on which to operate.
+///
+/// * `tag` - The tag to remove.
+///
+/// # Returns
+///
+/// There is no success return value.
+pub fn remove_tag(conn: &Connection, tag: String) -> Result<()> {
+    conn.execute_named_cached(
+        "DELETE FROM moz_tags
+         WHERE tag = :tag",
+        &[(":tag", &tag)],
+    )?;
+    Ok(())
+}
+
+/// Retrieves a list of URLs which have the specified tag.
+///
+/// # Arguments
+///
+/// * `conn` - A database connection on which to operate.
+///
+/// * `tag` - The tag to query.
+///
+/// # Returns
+///
+/// * A Vec<Url> with all URLs which have the tag, ordered by the frecency of
+/// the URLs.
+pub fn get_urls_with_tag(conn: &Connection, tag: String) -> Result<Vec<Url>> {
+    let tag = validate_tag(&tag)?;
+
+    let mut stmt = conn.prepare(
+        "SELECT p.url FROM moz_places p
+         JOIN moz_tags_relation r ON r.place_id = p.id
+         JOIN moz_tags t ON t.id = r.tag_id
+         WHERE t.tag = :tag
+         ORDER BY p.frecency",
+    )?;
+
+    let rows =
+        stmt.query_and_then_named(&[(":tag", &tag)], |row| row.get_checked::<_, String>("url"))?;
+    let mut urls = Vec::new();
+    for row in rows {
+        urls.push(Url::parse(&row?)?);
```

Comment:
rusqlite 0.17.0 has native URL support! (not that we can trivially change to it for this patch or anything, just wanted to note that)